### DISCUSSION TITLE
Material Varients/Instancing Feature?

### DISCUSSION THREAD

**Discussion by MangoDragonHub** [godotengine/godot-proposals#4540] [3D]
It would be cool if Godot 3 or 4 had a feature that allowed you to make one core shader material that is used on pretty much all meshes but you are able to create a variant of that core material with slight alterations. I'm not much of a graphics engineer but implementing this may result in performance improvements since it's relying on one main material shader than multiple individual shaders and with the new async shader compiler coming to 3.5, that will help out in shorting the compilation time as well,

**Comment by Zireael07** [comment]
I'm not sure what exactly you want, but "one main material shader" is what the ubershader approach in 3.x is based on, and StandardMaterial3D is already "one core shader material" that you can create a variant of.

**Comment by Calinou** [comment]
This is already supported with [per-instance uniforms in Godot 4.0](https://godotengine.org/article/godot-40-gets-global-and-instance-shader-uniforms). This can't be backported to Godot 3.x since it relies on functionality only found in Vulkan.

**Reply by MangoDragonHub** [reply]
Huh, that's interesting. Cause I'm more used to Unreal 4's method of allowing you to making one core/base Material and then you can make multiple material instances from that one material. Also, since you mention that is a Vulkan only feature, how come in softwares like Unreal allows to create something similar when they also have a OpenGL ES3 renderer as well? At least when you're developing for Android applications.

**Reply by Zireael07** [reply]
Unreal 4 uses ubershader approach and multiple material instances likely do what Godot GLES3 also can do, i.e. automatically duplicate() them under the hood.

**Reply by Calinou** [reply]
It *might* be possible to implement per-instance uniforms in GLES3, but it would likely be more limited and suffer from more edge cases. Therefore, it's unlikely to ever be implemented in `3.x` or in the GLES3 renderer in 4.x.
